### PR TITLE
feat: binary exponentiation

### PR DESCRIPTION
**Pull request by jay-tux**
#### Description of Change
Added binary exponentiation algorithm

#### Checklist

- [x] Added description of change
- [x] Added file name matches [File name guidelines](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/CONTRIBUTING.md#New-File-Name-guidelines)
- [x] Added tests and example, test must pass
- [x] Added documentation so that the program is self-explanatory and educational - [Doxygen guidelines](https://www.doxygen.nl/manual/docblocks.html)
- [x] PR title follows semantic [commit guidelines](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/CONTRIBUTING.md#Commit-Guidelines)
- [x] Search previous suggestions before making a new one, as yours may be a duplicate.
- [x] I acknowledge that all my contributions will be made under the project's license.

Notes: added binary exponentiation

<a href="https://gitpod.io/#https://github.com/TheAlgorithms/C-Plus-Plus/pull/1805"></a>

### CODE REVIEW

File: divide_and_conquer/binary_exponentiation.cpp [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+ * @file
+ * @brief Implementation of the [Binary Exponentiation
+ * (divide-and-conquer)](https://en.wikipedia.org/wiki/Exponentiation_by_squaring)
+ * algorithm
+ * @author [jay-tux](github.com/jay-tux)
+ * Computes base ^ exp by dividing exp by two while squaring base (reducing the
+ * amount of multiplications).
+ */
+
+#include <cassert>  /// for assert
```
Comment (Panquesito7) [inline divide_and_conquer/binary_exponentiation.cpp:11]:
```suggestion
#include <iostream> /// for IO operations
#include <cassert>  /// for assert
```

File: divide_and_conquer/binary_exponentiation.cpp [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+        for (int j = 0; j < 100; j++) {
+            // limit ranges to avoid overflow errors
+            base = std::rand() % 12;
+            exp = std::rand() % 12;
+            res = binaryExp(base, exp);
+            stdlib = static_cast<result_t>(
+                std::pow(static_cast<double>(base), static_cast<double>(exp)));
+            assert(res == stdlib);  // a^b = a^b
+        }
+    }
```
Comment (Panquesito7) [inline divide_and_conquer/binary_exponentiation.cpp:87]:
```suggestion
    }
    std::cout << "All tests have successfully passed!\n";
```

**Comment by Panquesito7** [comment]
A question: is the same algorithm as [**this**](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/math/binary_exponent.cpp) algorithm?
If so, we might have to rename this to `binary_exponentiation2.cpp`.

**Comment by jay-tux** [comment]
My bad. It seems that is the case.
Didn't check that directory as issue #1734 mentions it under divide and conquer.
Does this mean either of the two algorithms have to move to another directory?

**Comment by Panquesito7** [comment]
> My bad. It seems that is the case. Didn't check that directory as issue #1734 mentions it under divide and conquer. Does this mean either of the two algorithms have to move to another directory?

No, it's totally fine as-is 😄 Let us rename this file to `binary_exponentiation2.cpp`.

**Comment by jay-tux** [comment]
Renamed it!

File: divide_and_conquer/CMakeLists.txt [hunk 3]
```diff
@@ -0,0 +1,18 @@
+# If necessary, use the RELATIVE flag, otherwise each source file may be listed
+# with full pathname. RELATIVE may makes it easier to extract an executable name
+# automatically.
+file( GLOB APP_SOURCES RELATIVE ${CMAKE_CURRENT_SOURCE_DIR} binary_exponentiation2.cpp )
```
Comment (Panquesito7) [inline divide_and_conquer/CMakeLists.txt:4]:
```suggestion
file( GLOB APP_SOURCES RELATIVE ${CMAKE_CURRENT_SOURCE_DIR} *.cpp )
```

**Review by Panquesito7 — CHANGES_REQUESTED** [review]

**Review by Panquesito7 — DISMISSED** [review]
Awesome work! You've done amazing work, and it's been your
first contribution here! We hope you keep contributing! 😄👍🎉

After this PR is merged, you can ask for the
contributor role in our [**Discord server**](https://discord.gg/c7MnfGFGa6)! 🚀

**Review by amino19 — APPROVED** [review]

**Review by ayaankhan98 — APPROVED** [review]

**Review by ayaankhan98 — COMMENTED** [review]
Code looks good but windows CI is failing.

**Comment by Panquesito7** [comment]
Hmm, no idea why it's failing. Any ideas?

**Review by Panquesito7 — CHANGES_REQUESTED** [review]
Okay, I found the issue here Okay, I found the error here https://github.com/TheAlgorithms/C-Plus-Plus/runs/4096578335?check_suite_focus=true#step:4:250 could you please fix it? Let us know if you need any help! Thanks. 🙂

**Comment by jay-tux** [comment]
> Okay, I found the issue here Okay, I found the error here https://github.com/TheAlgorithms/C-Plus-Plus/runs/4096578335?check_suite_focus=true#step:4:250 could you please fix it? Let us know if you need any help! Thanks. slightly_smiling_face

I don't thing that's the issue. It points to a warning but it actually compiles to an executable afterwards, so that couldn't be the issue.

**Comment by Panquesito7** [comment]
> > Okay, I found the issue here Okay, I found the error here https://github.com/TheAlgorithms/C-Plus-Plus/runs/4096578335?check_suite_focus=true#step:4:250 could you please fix it? Let us know if you need any help! Thanks. slightly_smiling_face
> 
> I don't thing that's the issue. It points to a warning but it actually compiles to an executable afterwards, so that couldn't be the issue.

It is more of a warning, not an error, I think. But there could be a loss of data there, so please try to investigate. 🙂

**Comment by micwoj92** [comment]
This branch has conflicts that must be resolved.

**Pull request closed without merging**